sharktank/sharktank/layers/ffn_moe.py

```diff
@@ -0,0 +1,213 @@
+# Copyright 2024 Advanced Micro Devices, Inc
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from typing import Optional
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from .base import Theta, ThetaLayer, LinearLayer, RMSNormLayer
+
+__all__ = [
+    "MixtralBlockSparseTop2MLP",
+    "MixtralSparseMoeBlock",
+]
+
+
+class MixtralBlockSparseTop2MLP(ThetaLayer):
```

Comment:
Naming nit: This looks just like a normal FFN network. I probably should have made this for the base model. We don't need to call this MixtralBlockSparseTop2MLP. It could just be LlmFfnMLP (or something like that). I'll need to do work on this module to shard/distribute it, and would like to make it common.

Reply:
Thanks for taking a look Stella. Yes, it's a common FFN and I will extract it into it's own file. Once I wrap Mixtral, will have a PR to add the FFN to existing Llama model too.